README.md
# Palmtrie: A Ternary Key Matching Algorithm for IP Packet Filtering Rules

## About

This is the reference implementation of "Palmtrie: A Ternary Key Matching
Algorithm for IP Packet Filtering Rules," to appear in CoNEXT 2020.


## Author

Hirochika Asai


## License

The use of this software is limited to education, research, and evaluation
purposes only.  Commercial use is strictly prohibited.  For all other uses,
contact the author(s).

## Getting started

This software uses automake/autoconf.  Run `./autogen.sh` to generate a
`configure` script.  After generating the `configure` script or if you unarchive
a source code archive with the `configure` script, run `./configure` and `make`
to compile the software.

### Basic tests and microbenchmarking

To run basic tests and microbenchmarks, the `make test` command can be used.

    $ make test
    /usr/bin/make  all-recursive
    make[2]: Nothing to be done for `all-am'.
    Testing all...
    ./palmtrie_test_basic
    basic test (SORTED_LIST): passed
    basic test (BASIC): passed
    basic test (DEFAULT): passed
    basic test (PLUS): passed
    microbenchmarking (SORTED_LIST): ..............(26.088413 sec: 0.000 Mlookup/sec)passed
    microbenchmarking (BASIC): ..............(9.045084 sec: 1.855 Mlookup/sec)passed
    microbenchmarking (DEFAULT): ..............(4.566560 sec: 3.674 Mlookup/sec)passed
    microbenchmarking (PLUS): ..............(2.386559 sec: 7.030 Mlookup/sec)passed
    cross check (BASIC,DEFAULT): ....................................................................passed
    cross check (SORTED_LIST,BASIC): ....................................................................passed
    cross check for ACL (BASIC,DEFAULT): .................passed
    cross check for ACL reverse order scanning (BASIC,DEFAULT): ..........passed
    cross check for ACL reverse order scanning (BASIC,PLUS): ..........passed
    cross check for ACL (SORTED_LIST,BASIC): .................passed
    cross check for ACL reverse order scanning (SORTED_LIST,BASIC): ..........passed

### Evaluation

The compiled programs named `palmtrie_eval_lpm` and `palmtrie_eval_acl` run
performance evaluation of Palmtrie variants as well as the conventional sorted
list for longest prefix matching and ternary matching for ACLs, respectively.

The `ptcam_eval_lpm` program takes two arguments: 1) routing table and 2) type
of the data structure (see the initialization section below for the details);
sl) SORTED_LIST, tpt) BASIC, mtpt) DEFAULT, and popmtpt) PLUS.  A sammple
routing table is found at `tests/linx-rib.20141217.0000-p46.txt` in this source
code directory.

The `ptcam_eval_acl` program takes two arguments: 1) ternary matching table and
2) type of the data structure and traffic pattern.  The second argument can be
separated by - (dash) and its prefix and suffix represents the type of the data
structure and the traffic pattern, respectively.  The type is same as the
description for the `ptcam_eval_lpm` program above.  The traffic pattern are
either of rand) random source address, random destination address in the range
of 10.0.0.0/8, and random source and destination port numbers, sfl) traffic
pattern specified by the `tests/traffic.sfl2` file, ross) reverse-byte order
scanning.  Examples of ternary matching tables are found at
`tests/acl-0001.tcam` and `tests/acl-0002.tcam`.

The following toolset is used to convert an ACL ruleset to a ternary matching
table and generate a traffic pattern file: https://github.com/drpnd/acl

## APIs

### Initialization

    NAME
         palmtrie_init -- initialize a palmtrie control data structure

    SYNOPSIS
         struct palmtrie *
         palmtrie_init(struct palmtrie *palmtrie, enum palmtrie_type type);

    DESCRIPTION
         The palmtrie_init() function initializes a palmtrie control data
         structure specified by the palmtrie argument with a type parameter.

         If the palmtrie argument is NULL, a new data structure is allocated and
         initialized.  Otherwise, the data structure specified by the palmtrie
         argument is initialized.

         The type parameter specifies the type of the palmtrie.  The valid type
         are the followings:

         o PALMTRIE_SORTED_LIST: A reference type to implement ternary matching
           using a sorted list, not Palmtrie.

         o PALMTRIE_BASIC: A type to represent Palmtrie (basic) that implements
           a single-bit stride trie.

         o PALMTRIE_DEFAULT: A type to represent Palmtrie that implements
           a multibit stride extension.

         o PALMTRIE_PLUS: A type to represent Palmtrie+ that implements an
           optimization technique using the population count instruction.
           The Palmtrie data structure requires to call the palmtrie_commit()
           function after updating the data structure.

    RETURN VALUES
         Upon successful completion, the palmtrie_init() function returns the
         pointer to the initialized palmtrie data structure.  Otherwise, it
         returns a NULL value and set errno.  If a non-NULL poptrie argument is
         specified, the returned value shall be the original value of the
         poptrie argument if successful, or a NULL value otherwise.

### Addition

    NAME
         palmtrie_add_data -- add an entry to the palmtrie data structure

    SYNOPSIS
         int
         palmtrie_add_data(struct palmtrie *palmtrie, addr_t addr, addr_t mask,
                           int priority, uint64_t data);

    DESCRIPTION
         The palmtrie_add_data() function adds an entry with a ternary key data
         specified by a pair of addr and mask arguments, the priority, and the
         data into the trie specified by the palmtrie argument.


    RETURN VALUES
         On successful, the palmtrie_add_data() function returns a value of 0.
         Otherwise, they return a value of -1.

### Lookup

    NAME
         palmtrie_lookup -- obtain a 64-bit data by looking up the entry
         corresponding the specified key from the palmtrie data structure

    SYNOPSIS
         uint64_t
         palmtrie_lookup(struct palmtrie *palmtrie, addr_t addr);

    DESCRIPTION
         The palmtrie_lookup() function looks up an entry corresponding to the
         key specified by the addr argument.

    RETURN VALUES
         The palmtrie_lookup() function returns a 64-bit data corresponding to
         the addr argument.  If no matching entry is found, a zero value is
         returned.



ChangeLog
See the commit log.



tests/acl.c
/*_
 * Copyright (c) 2018-2020 Hirochika Asai <asai@jar.jp>
 * All rights reserved.
 */

#include "../palmtrie.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>

#include <alloca.h>
void
reverse(char *s)
{
    char *t;
    int l;
    int i;
    l = strlen(s);
    t = alloca(l + 1);
    for ( i = 0; i < l; i++ ) {
        t[i] = s[l - i - 1];
    }
    t[l] = 0;

    strcpy(s, t);
}

/* Macro for testing */
#define TEST_FUNC(str, func, ret)                \
    do {                                         \
        printf("%s: ", str);                     \
        if ( 0 == func() ) {                     \
            printf("passed");                    \
        } else {                                 \
            printf("failed");                    \
            ret = -1;                            \
        }                                        \
        printf("\n");                            \
    } while ( 0 )

#define TEST_PROGRESS()                              \
    do {                                             \
        printf(".");                                 \
        fflush(stdout);                              \
    } while ( 0 )

/*
 * Get current time at the microsecond granularity
 */
double
getmicrotime(void)
{
    struct timeval tv;
    double microsec;

    if ( 0 != gettimeofday(&tv, NULL) ) {
        return 0.0;
    }

    microsec = (double)tv.tv_sec + (1.0 * tv.tv_usec / 1000000);

    return microsec;
}

/*
 * Xorshift
 */
static __inline__ u32
xor128(void)
{
    static u32 x = 123456789;
    static u32 y = 362436069;
    static u32 z = 521288629;
    static u32 w = 88675123;
    u32 t;

    t = x ^ (x<<11);
    x = y;
    y = z;
    z = w;
    return w = (w ^ (w>>19)) ^ (t ^ (t >> 8));
}

static __inline__ int
hex2bin(char c)
{
    if ( c >= '0' && c <= '9' ) {
        return c - '0';
    } else if ( c >= 'a' && c <= 'f' ) {
        return c - 'a' + 10;
    } else if ( c >= 'A' && c <= 'F' ) {
        return c - 'A' + 10;
    } else {
        return 0;
    }
}

/*
 * Performance test
 */
static int
test_microperf(enum palmtrie_type type, const char *fname, long long nr)
{
    struct palmtrie palmtrie;
    FILE *fp;
    char buf[4096];
    char data0[1024];
    char data1[1024];
    int priority;
    int action;
    int ret;
    addr_t addr = {0, {0, 0, 0, 0, 0, 0, 0, 0}};
    addr_t mask = {0, {0, 0, 0, 0, 0, 0, 0, 0}};
    u64 d;
    long long i;
    ssize_t k;
    addr_t tmp = {0, {0, 0, 0, 0, 0, 0, 0, 0}};
    double t0;
    double t1;
    double delta;
    u64 x;

    /* Initialize */
    palmtrie_init(&palmtrie, type);

    /* Load from the linx file */
    fp = fopen(fname, "r");
    if ( NULL == fp ) {
        return -1;
    }

    /* Load the full route */
    i = 0;
    while ( !feof(fp) ) {
        if ( !fgets(buf, sizeof(buf), fp) ) {
            continue;
        }
        ret = sscanf(buf, "%1000s %1000s %d %d", data0, data1, &priority,
                     &action);
        if ( ret < 0 ) {
            return -1;
        }
        if ( strlen(data0) != 32 || strlen(data1) != 32 ) {
            /* Length mismatch */
            return -1;
        }
        reverse(data0);
        reverse(data1);
        memset(&addr, 0, sizeof(addr_t));
        memset(&mask, 0, sizeof(addr_t));
        for ( k = 0; k < (ssize_t)strlen(data0); k++ ) {
            d = hex2bin(data0[k]);
            addr.a[k >> 4] |= d << ((k & 0xf) << 2);
            d = hex2bin(data1[k]);
            mask.a[k >> 4] |= d << ((k & 0xf) << 2);
        }
        ret = palmtrie_add_data(&palmtrie, addr, mask, priority, action);
        if ( ret < 0 ) {
            return -1;
        }
        if ( 0 == i % 100000 ) {
            TEST_PROGRESS();
        }
        i++;
    }
    ret = palmtrie_commit(&palmtrie);
    if ( ret < 0 ) {
        return -1;
    }

    /* Close */
    fclose(fp);

    /* Benchmark */
    x = 0;
    t0 = getmicrotime();
    for ( i = 0; i < nr; i++ ) {
        if ( 0 == i % (nr / 8) ) {
            TEST_PROGRESS();
        }
        tmp.a[0] = xor128();
        tmp.a[1] = xor128();
        x ^= palmtrie_lookup(&palmtrie, tmp);
    }
    t1 = getmicrotime();
    (void)x;
    delta = t1 - t0;
    printf("(%.6lf sec: %.3lf Mlookup/sec)", delta, i / delta / 1000 / 1000);

    return 0;
}
static int
test_microperf_sl(void)
{
    return test_microperf(PALMTRIE_SORTED_LIST, "tests/acl-0002.tcam",
                          0x100000ULL);
}
static int
test_microperf_tpt(void)
{
    return test_microperf(PALMTRIE_BASIC, "tests/acl-0002.tcam",
                          0x1000000ULL);
}
static int
test_microperf_mtpt(void)
{
    return test_microperf(PALMTRIE_DEFAULT,
                          "tests/acl-0002.tcam", 0x1000000ULL);
}
static int
test_microperf_popmtpt(void)
{
    return test_microperf(PALMTRIE_PLUS,
                          "tests/acl-0002.tcam", 0x1000000ULL);
}

/*
 * Performance test
 */
static int
test_rossperf(enum palmtrie_type type, const char *fname, const char *tfname,
              long long nr)
{
    struct palmtrie palmtrie;
    FILE *fp;
    char buf[4096];
    char data0[1024];
    char data1[1024];
    int priority;
    int action;
    int ret;
    addr_t addr = {0, {0, 0, 0, 0, 0, 0, 0, 0}};
    addr_t mask = {0, {0, 0, 0, 0, 0, 0, 0, 0}};
    u64 d;
    long long i;
    ssize_t k;
    addr_t tmp = {0, {0, 0, 0, 0, 0, 0, 0, 0}};
    double t0;
    double t1;
    double delta;
    u64 x;
    addr_t *pattern;
    int npkt;
    long long j;

    /* Initialize */
    palmtrie_init(&palmtrie, type);

    /* Load TCAM file */
    fp = fopen(fname, "r");
    if ( NULL == fp ) {
        return -1;
    }

    /* Load the full route */
    i = 0;
    while ( !feof(fp) ) {
        if ( !fgets(buf, sizeof(buf), fp) ) {
            continue;
        }
        ret = sscanf(buf, "%1000s %1000s %d %d", data0, data1, &priority,
                     &action);
        if ( ret < 0 ) {
            return -1;
        }
        if ( strlen(data0) != 32 || strlen(data1) != 32 ) {
            /* Length mismatch */
            return -1;
        }
        reverse(data0);
        reverse(data1);
        memset(&addr, 0, sizeof(addr_t));
        memset(&mask, 0, sizeof(addr_t));
        for ( k = 0; k < (ssize_t)strlen(data0); k++ ) {
            d = hex2bin(data0[k]);
            addr.a[k >> 4] |= d << ((k & 0xf) << 2);
            d = hex2bin(data1[k]);
            mask.a[k >> 4] |= d << ((k & 0xf) << 2);
        }
        ret = palmtrie_add_data(&palmtrie, addr, mask, priority, action);
        if ( ret < 0 ) {
            return -1;
        }
        if ( 0 == i % 100000 ) {
            TEST_PROGRESS();
        }
        i++;
    }
    ret = palmtrie_commit(&palmtrie);
    if ( ret < 0 ) {
        return -1;
    }

    /* Close */
    fclose(fp);

    /* Traffic pattern */
    fp = fopen(tfname, "r");
    if ( NULL == fp ) {
        return -1;
    }

    /* Load the full route */
    i = 0;
    pattern = malloc(sizeof(addr_t) * 65536 * 256);
    while ( !feof(fp) ) {
        if ( !fgets(buf, sizeof(buf), fp) ) {
            continue;
        }
        ret = sscanf(buf, "%1000s", data0);
        if ( ret < 0 ) {
            return -1;
        }
        reverse(data0);
        if ( strlen(data0) != 32 ) {
            /* Length mismatch */
            return -1;
        }
        memset(&tmp, 0, sizeof(addr_t));
        for ( k = 0; k < (ssize_t)strlen(data0); k++ ) {
            d = hex2bin(data0[k]);
            tmp.a[k >> 4] |= d << ((k & 0xf) << 2);
        }
        (void)memcpy(&pattern[i], &tmp, sizeof(addr_t));
        if ( 0 == i % 1000000 ) {
            TEST_PROGRESS();
        }
        i++;
        if ( i >= 65536 * 256 ) {
            break;
        }
    }
    npkt = i;

    /* Close */
    fclose(fp);

    /* Benchmark */
    x = 0;
    t0 = getmicrotime();
    for ( i = 0, j = 0; i < nr; i++ ) {
        if ( 0 == i % (nr / 8) ) {
            TEST_PROGRESS();
        }
        x ^= palmtrie_lookup(&palmtrie, pattern[j]);
        j++;
        if ( j >= npkt ) {
            j = 0;
        }
    }
    t1 = getmicrotime();
    (void)x;
    delta = t1 - t0;
    printf("(%.6lf sec: %.3lf Mlookup/sec)", delta, i / delta / 1000 / 1000);

    return 0;
}
static int
test_rossperf_sl(void)
{
    return test_rossperf(PALMTRIE_SORTED_LIST, "tests/acl-0001.tcam",
                         "tests/acl-0001.ross", 0x100000ULL);
}
static int
test_rossperf_tpt(void)
{
    return test_rossperf(PALMTRIE_BASIC, "tests/acl-0001.tcam",
                         "tests/acl-0001.ross",
                         0x1000000ULL);
}
static int
test_rossperf_mtpt(void)
{
    return test_rossperf(PALMTRIE_DEFAULT, "tests/acl-0001.tcam",
                         "tests/acl-0001.ross", 0x1000000ULL);
}
static int
test_rossperf_popmtpt(void)
{
    return test_rossperf(PALMTRIE_PLUS,
                         "tests/acl-0001.tcam", "tests/acl-0001.ross",
                         0x1000000ULL);
}
static int
test_rossperf2_sl(void)
{
    return test_rossperf(PALMTRIE_SORTED_LIST, "tests/acl-0002.tcam",
                         "tests/acl-0001.ross", 0x100000ULL);
}
static int
test_rossperf2_tpt(void)
{
    return test_rossperf(PALMTRIE_BASIC, "tests/acl-0002.tcam",
                         "tests/acl-0001.ross", 0x1000000ULL);
}
static int
test_rossperf2_mtpt(void)
{
    return test_rossperf(PALMTRIE_DEFAULT, "tests/acl-0002.tcam",
                         "tests/acl-0001.ross", 0x1000000ULL);
}
static int
test_rossperf2_popmtpt(void)
{
    return test_rossperf(PALMTRIE_PLUS,
                         "tests/acl-0002.tcam", "tests/acl-0001.ross",
                         0x1000000ULL);
}
static int
test_rossperf3_sl(void)
{
    return test_rossperf(PALMTRIE_SORTED_LIST, "tests/acl-1000.tcam",
                         "tests/acl-1000.ross", 0x100000ULL);
}
static int
test_rossperf3_tpt(void)
{
    return test_rossperf(PALMTRIE_BASIC, "tests/acl-1000.tcam",
                         "tests/acl-1000.ross", 0x1000000ULL);
}
static int
test_rossperf3_mtpt(void)
{
    return test_rossperf(PALMTRIE_DEFAULT, "tests/acl-1000.tcam",
                         "tests/acl-1000.ross", 0x1000000ULL);
}
static int
test_rossperf3_popmtpt(void)
{
    return test_rossperf(PALMTRIE_PLUS,
                         "tests/acl-1000.tcam", "tests/acl-1000.ross",
                         0x1000000ULL);
}

/*
 * Main routine for the basic test
 */
int
main(int argc, const char *const argv[])
{
    int ret;
    int flags;
    int i;

    ret = 0;

    flags = 0;
    if ( argc < 2 ) {
        /* Test all */
        flags = 0xff;
    } else {
        for ( i = 1; i < argc; i++ ) {
            if ( 0 == strcmp("microperf", argv[i]) ) {
                flags |= 1 << 1;
            } else if ( 0 == strcmp("rossperf1", argv[i]) ) {
                flags |= 1 << 2;
            } else if ( 0 == strcmp("rossperf2", argv[i]) ) {
                flags |= 1 << 3;
            } else if ( 0 == strcmp("rossperf3", argv[i]) ) {
                flags |= 1 << 4;
            }
        }
    }

    /* Run tests */
    if ( flags & (1 << 1) ) {
        /* Micro performance measurement */
        TEST_FUNC("microperf (sl)", test_microperf_sl, ret);
        TEST_FUNC("microperf (tpt)", test_microperf_tpt, ret);
        TEST_FUNC("microperf (mtpt)", test_microperf_mtpt, ret);
        TEST_FUNC("microperf (popmtpt)", test_microperf_popmtpt, ret);
    }
    if ( flags & (1 << 2) ) {
        /* Micro performance measurement */
        TEST_FUNC("rossperf1 (sl)", test_rossperf_sl, ret);
        TEST_FUNC("rossperf1 (tpt)", test_rossperf_tpt, ret);
        TEST_FUNC("rossperf1 (mtpt)", test_rossperf_mtpt, ret);
        TEST_FUNC("rossperf1 (popmtpt)", test_rossperf_popmtpt, ret);
    }
    if ( flags & (1 << 3) ) {
        /* Micro performance measurement */
        TEST_FUNC("rossperf2 (sl)", test_rossperf2_sl, ret);
        TEST_FUNC("rossperf2 (tpt)", test_rossperf2_tpt, ret);
        TEST_FUNC("rossperf2 (mtpt)", test_rossperf2_mtpt, ret);
        TEST_FUNC("rossperf2 (popmtpt)", test_rossperf2_popmtpt, ret);
    }
    if ( flags & (1 << 4) ) {
        /* Micro performance measurement */
        //TEST_FUNC("rossperf3 (sl)", test_rossperf3_sl, ret);
        TEST_FUNC("rossperf3 (tpt)", test_rossperf3_tpt, ret);
        TEST_FUNC("rossperf3 (mtpt)", test_rossperf3_mtpt, ret);
        TEST_FUNC("rossperf3 (popmtpt)", test_rossperf3_popmtpt, ret);
    }

    return ret;
}

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */


autogen.sh
#!/bin/sh

(autoreconf --version) < /dev/null > /dev/null 2>&1 || {
  (autoconf --version) < /dev/null > /dev/null 2>&1 || {
    echo
    echo "**Error**: You must have the GNU Build System (autoconf, automake, "
    echo "libtool, etc) to update the build system.  Download the appropriate"
    echo "packages for your distribution, or get the source tar balls from"
    echo "ftp://ftp.gnu.org/pub/gnu/."
    exit 1
  }
  echo
  echo "**Error**: Your version of autoconf is too old (you need at least 2.57)"
  echo "to update the build system.  Download the appropriate updated package"
  echo "for your distribution, or get the source tar ball from"
  echo "ftp://ftp.gnu.org/pub/gnu/."
  exit 1
}

echo Running autoreconf --verbose --install --force
autoreconf --verbose --install --force

echo Removing autom4te.cache
rm -rf autom4te.cache


palmtrie.h
/*_
 * Copyright (c) 2015-2020 Hirochika Asai <asai@jar.jp>
 * All rights reserved.
 */

#ifndef _PALMTRIE_H
#define _PALMTRIE_H

#include <stdint.h>
#include <immintrin.h>

typedef uint16_t u16;
typedef uint32_t u32;
typedef uint64_t u64;

/* Typical TCAM key-width: 40, 80, 160, 320, 480, 640 bits */

/* 512-bit addressing */
#if defined(PALMTRIE_SHORT) && PALMTRIE_SHORT
typedef struct { u32 g; u64 a[2]; } __attribute__ ((packed)) addr_t;
#define PALMTRIE_ADDR_BITS     480
#define PALMTRIE_ADDR_ZERO     {0, {0, 0}}

#define ADDR_MASK(addr, mask) do {              \
        (addr).a[0] &= ~(mask).a[0];            \
        (addr).a[1] &= ~(mask).a[1];            \
    } while ( 0 )

#define ADDR_MASK_CMP(addr0, mask0, addr1, mask1)                       \
    (((addr0).a[0] & ~(mask0).a[0]) == ((addr1).a[0] & ~(mask1).a[0])   \
     && ((addr0).a[1] & ~(mask0).a[1]) == ((addr1).a[1] & ~(mask1).a[1]))

#define ADDR_MASK_CMP2(addr0, mask0, addr1)                             \
    (((addr0).a[0] & ~(mask0).a[0]) == ((addr1).a[0])                   \
     && ((addr0).a[1] & ~(mask0).a[1]) == ((addr1).a[1]))

#define ADDR_CMP(x, y)                                                  \
    ((x).a[0] == (y).a[0] && (x).a[1] == (y).a[1])

#else

typedef struct { u32 g; u64 a[8]; } __attribute__ ((packed)) addr_t;
#define PALMTRIE_ADDR_BITS     480
#define PALMTRIE_ADDR_ZERO     {0, {0, 0, 0, 0, 0, 0, 0, 0}}

#define ADDR_MASK(addr, mask) do {              \
        (addr).a[0] &= ~(mask).a[0];            \
        (addr).a[1] &= ~(mask).a[1];            \
        (addr).a[2] &= ~(mask).a[2];            \
        (addr).a[3] &= ~(mask).a[3];            \
        (addr).a[4] &= ~(mask).a[4];            \
        (addr).a[5] &= ~(mask).a[5];            \
        (addr).a[6] &= ~(mask).a[6];            \
        (addr).a[7] &= ~(mask).a[7];            \
    } while ( 0 )

#define ADDR_MASK_CMP(addr0, mask0, addr1, mask1)                       \
    (((addr0).a[0] & ~(mask0).a[0]) == ((addr1).a[0] & ~(mask1).a[0])   \
     && ((addr0).a[1] & ~(mask0).a[1]) == ((addr1).a[1] & ~(mask1).a[1]) \
     && ((addr0).a[2] & ~(mask0).a[2]) == ((addr1).a[2] & ~(mask1).a[2]) \
     && ((addr0).a[3] & ~(mask0).a[3]) == ((addr1).a[3] & ~(mask1).a[3]) \
     && ((addr0).a[4] & ~(mask0).a[4]) == ((addr1).a[4] & ~(mask1).a[4]) \
     && ((addr0).a[5] & ~(mask0).a[5]) == ((addr1).a[5] & ~(mask1).a[5]) \
     && ((addr0).a[6] & ~(mask0).a[6]) == ((addr1).a[6] & ~(mask1).a[6]) \
     && ((addr0).a[7] & ~(mask0).a[7]) == ((addr1).a[7] & ~(mask1).a[7]))

#define ADDR_MASK_CMP2(addr0, mask0, addr1)                             \
    (((addr0).a[0] & ~(mask0).a[0]) == ((addr1).a[0])                   \
     && ((addr0).a[1] & ~(mask0).a[1]) == ((addr1).a[1])                \
     && ((addr0).a[2] & ~(mask0).a[2]) == ((addr1).a[2])                \
     && ((addr0).a[3] & ~(mask0).a[3]) == ((addr1).a[3])                \
     && ((addr0).a[4] & ~(mask0).a[4]) == ((addr1).a[4])                \
     && ((addr0).a[5] & ~(mask0).a[5]) == ((addr1).a[5])                \
     && ((addr0).a[6] & ~(mask0).a[6]) == ((addr1).a[6])                \
     && ((addr0).a[7] & ~(mask0).a[7]) == ((addr1).a[7]))

#define ADDR_CMP(x, y)                                                  \
    ((x).a[0] == (y).a[0] && (x).a[1] == (y).a[1] && (x).a[2] == (y).a[2] \
     && (x).a[3] == (y).a[3] && (x).a[4] == (y).a[4] && (x).a[5] == (y).a[5] \
     && (x).a[6] == (y).a[6] && (x).a[7] == (y).a[7])

#endif


#ifndef PALMTRIE_MTPT_STRIDE
#define PALMTRIE_MTPT_STRIDE  8
#endif

#ifndef PALMTRIE_PRIORITY_SKIP
#define PALMTRIE_PRIORITY_SKIP 1
#endif

#ifndef PALMTRIE_EXACTMATCH_FIRST
#define PALMTRIE_EXACTMATCH_FIRST 0
#endif


static __inline__ int
ADDR_PREFIX_CMP(addr_t a0, addr_t m0, addr_t a1, addr_t m1, int plen, int msb)
{
    int i;

    i = (plen >> 6);
    if ( ((a0.a[i] & ~(m0.a[i])) >> (plen & 0x3f))
         != ((a1.a[i] & ~(m1.a[i])) >> (plen & 0x3f)) ) {
        return 0;
    }
    i++;

    for ( ; i <= (msb >> 6); i++ ) {
        if ( (a0.a[i] & ~(m0).a[i]) != (a1.a[i] & ~(m1).a[i]) ) {
            return 0;
        }
    }

    return 1;
}

#define EXTRACT(addr, bit)      (((addr).a[(bit) >> 6] >> ((bit) & 0x3f)) & 1)

#define EXTRACTN(addr, bit, n)                                      \
    ((*(uint32_t *)((void *)&(addr).g + (((bit) + 32) >> 3))        \
      >> (((bit) + 0) & 0x7)) & ((1 << (n)) - 1))


#define BTS(addr, bit) do {                                 \
        (addr).a[(bit) >> 6] |= (1ULL << ((bit) & 0x3f));   \
    } while ( 0 )
#define BTC(addr, bit) do {                             \
        (addr).a[(bit) >> 6] &= ~(1ULL << ((bit) & 0x3f));  \
    } while ( 0 )

#define BACKTRACK_SKIP(n)   ((n).p & 0x7)
#define BACKTRACK_NODE(n)                                       \
    (struct palmtrie_mtpt_node *)((n).p & 0xfffffffffffffff8ULL)


/*
 * Type-packed union pointer
 */
#ifndef __LP64__
#error "Must be LP64."
#endif
#define TPUP_PTR(p)     (void *)((uint64_t)(p) & 0xfffffffffffffff8ULL)
#define TPUP_TYPE(p)    (int)((uint64_t)(p) & 0x7)
#define TPUP_SET(p, t)  (void *)((uint64_t)(p) | (t))


/*
 * Types of the software TCAM implementation
 */
enum palmtrie_type {
    PALMTRIE_SORTED_LIST,
    PALMTRIE_BASIC,
    PALMTRIE_DEFAULT,
    PALMTRIE_PLUS,
};

/*
 * An entry of the sorted list
 */
struct palmtrie_sorted_list_entry {
    addr_t addr;
    addr_t mask;
    int priority;
    void *data;
    struct palmtrie_sorted_list_entry *next;
};

/*
 * Sorted list
 */
struct palmtrie_sorted_list {
    struct palmtrie_sorted_list_entry *head;
};

/*
 * Node of ternary PATRICIA trie
 */
struct palmtrie_tpt_node {
    int bit;

    addr_t addr;
    addr_t mask;

#if PALMTRIE_PRIORITY_SKIP
    int max_priority;
#endif
    int priority;
    void *data;
    struct palmtrie_tpt_node *left;
    struct palmtrie_tpt_node *center;
    struct palmtrie_tpt_node *right;
};

/*
 * Ternary PATRICIA trie
 */
struct palmtrie_tpt {
    struct palmtrie_tpt_node *root;
};

/*
 * Node of multiway ternary PATRICIA trie
 * Types in the least significant 3 bits:
 *   Data node: 0
 *   Backtrack node: 1
 */
struct palmtrie_mtpt_node_backtrack {
    struct palmtrie_mtpt_node_data *backtrack;
    void *searchback;
};
struct palmtrie_mtpt_node_data {
    int bit;
    addr_t addr;
    addr_t mask;
    int priority;
#if PALMTRIE_PRIORITY_SKIP
    int max_priority;
#endif
    void *data;

    /* Descendent nodes */
    struct palmtrie_mtpt_node_data *children[1 << PALMTRIE_MTPT_STRIDE];
    /* Searchback nodes */
    struct palmtrie_mtpt_node_data *ternaries[(1 << PALMTRIE_MTPT_STRIDE) - 1];
};

/*
 * Multiway ternary PATRICIA trie
 */
struct palmtrie_mtpt {
    struct palmtrie_mtpt_node_data *root;
};

struct palmtrie_popmtpt_leaf
{
    int32_t priority;
    addr_t addr;
    addr_t mask;
    void *data;
};
#define PALMTRIE_STRIDE_OPT 1
struct palmtrie_popmtpt_node {
    int16_t bit;
    union {
        struct {
#if PALMTRIE_PRIORITY_SKIP
            int32_t max_priority;
#endif
#if PALMTRIE_MTPT_STRIDE == 8 && PALMTRIE_STRIDE_OPT
            uint32_t cbase;
            uint32_t tbase;
            uint8_t children[4];
            uint8_t ternaries[4];
            uint64_t bitmap_t[4];
            uint64_t bitmap_c[4];
#elif PALMTRIE_MTPT_STRIDE >= 6
            uint32_t children[((1 << PALMTRIE_MTPT_STRIDE) >> 6)];
            uint32_t ternaries[((1 << PALMTRIE_MTPT_STRIDE) >> 6)];
            uint64_t bitmap_t[((1 << PALMTRIE_MTPT_STRIDE) >> 6)];
            uint64_t bitmap_c[((1 << PALMTRIE_MTPT_STRIDE) >> 6)];
#else
            uint32_t children[1];
            uint32_t ternaries[1];
            uint64_t bitmap_t[1];
            uint64_t bitmap_c[1];
#endif
        } inode;
        struct {
            int32_t priority;
            addr_t addr;
            addr_t mask;
            void *data;
        } leaf;
    } u;
};
struct palmtrie_popmtpt {
    uint32_t root;
    struct {
        int nr;
        int used;
        struct palmtrie_popmtpt_node *ptr;
    } nodes;
    struct palmtrie_mtpt mtpt;
};

/*
 * Software TCAM
 */
struct palmtrie {
    union {
        struct palmtrie_sorted_list sl;
        struct palmtrie_tpt tpt;
        struct palmtrie_mtpt mtpt;
        struct palmtrie_popmtpt popmtpt;
    } u;
    enum palmtrie_type type;
    int allocated;
};

/* Prototype declarations */
struct palmtrie * palmtrie_init(struct palmtrie *, enum palmtrie_type);
int palmtrie_add_data(struct palmtrie *, addr_t, addr_t, int, u64);
u64 palmtrie_lookup(struct palmtrie *, addr_t);
int palmtrie_commit(struct palmtrie *);

/* in sl.c */
int palmtrie_sl_add(struct palmtrie *, addr_t, addr_t, int, void *);
void * palmtrie_sl_lookup(struct palmtrie *, addr_t);
int palmtrie_sl_release(struct palmtrie *);

/* in tpt.c */
int palmtrie_tpt_add(struct palmtrie *, addr_t, addr_t, int, void *);
void * palmtrie_tpt_lookup(struct palmtrie *, addr_t);
int palmtrie_tpt_release(struct palmtrie *);

/* in mtpt.c */
int palmtrie_mtpt_add(struct palmtrie_mtpt *, addr_t, addr_t, int, void *);
void * palmtrie_mtpt_lookup(struct palmtrie *, addr_t);
int palmtrie_mtpt_release(struct palmtrie *);

/* in popmtpt.c */
void * palmtrie_popmtpt_lookup(struct palmtrie_popmtpt *, addr_t);
int
palmtrie_popmtpt_add(struct palmtrie_popmtpt *, addr_t, addr_t, int, void *);
int palmtrie_popmtpt_commit(struct palmtrie_popmtpt *);

#endif

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */
